MedtrumKit/PumpManager/BluetoothManager.swift

```diff
@@ -0,0 +1,105 @@
+//
+//  BluetoothManager.swift
+//  MedtrumKit
+//
+//  Created by Bastiaan Verhaar on 25/02/2025.
+//
+
+import CoreBluetooth
+
+class BluetoothManager: NSObject, CBCentralManagerDelegate {
+    public var pumpManager: MedtrumPumpManager?
+    
+    let log = MedtrumLogger(category: "BluetoothManager")
+    
+    var manager: CBCentralManager!
+    let managerQueue = DispatchQueue(label: "com.nightscout.MedtrumKit.bluetoothManagerQueue", qos: .unspecified)
+    
+    private var peripheralManager: PeripheralManager?
+    
+    var scanCompletion: ((MedtrumScanResult) -> Void)?
+    var connectCompletion: ((MedtrumConnectResult) -> Void)?
+    
+    override init() {
+        super.init()
+        
+        managerQueue.sync {
+            self.manager = CBCentralManager(delegate: self, queue: managerQueue)
+        }
+    }
+    
+    func startScan(_ completion: @escaping (_ result: MedtrumScanResult) -> Void) {
+        guard manager.state == .poweredOn else {
+            completion(.failure(error: .invalidBluetoothState(state: manager.state)))
+            return
+        }
+
+        guard !manager.isScanning else {
+            completion(.failure(error: .alreadyScanning))
+            return
+        }
+
+        scanCompletion = completion
+        manager.scanForPeripherals(withServices: [])
+        
+        log.info("Started scanning")
+    }
+    
+    func connect(peripheral: CBPeripheral, _ completion: @escaping (MedtrumConnectResult) -> Void) {
+        if manager.isScanning {
+            manager.stopScan()
+            scanCompletion = nil
+        }
+        
+        log.info("Connecting to \(peripheral)")
+        
+        connectCompletion = completion
+        manager.connect(peripheral)
+    }
+}
+
+extension BluetoothManager {
+    func centralManagerDidUpdateState(_ central: CBCentralManager) {
+        log.info("\(String(describing: central.state.rawValue))")
+    }
+
+    func centralManager(_: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String: Any], rssi _: NSNumber) {
+        guard let deviceName = peripheral.name, !deviceName.isEmpty else {
+            return
+        }
+        
+        // TODO: Need to validate if the device name is always MT
```

Comment:
Yes, also ensure it's a perfect match, pump boots up in bootloader mode and advertises as "MTBT" you don't want o connect to that

Reply:
Another solution is to filter bootloader out by device type:
https://github.com/jbr7rr/insuBox/blob/8f217de4df784e86a49ee380fc07968353325e9b/lib/pump/medtrum_bt/comm/PumpScanner.cpp#L68